jmxtrans-output/jmxtrans-output-cloudwatch/src/main/java/com/googlecode/jmxtrans/model/output/CloudWatchWriter.java

```diff
@@ -22,152 +22,123 @@
+				this.dimensions = dimensions;
+				this.typeNames = typeNames;
 		}
 
 		@Override
 		public void doWrite(Server server, Query query, Iterable<Result> results) throws Exception {
-			PutMetricDataRequest metricDataRequest = new PutMetricDataRequest();
-			metricDataRequest.setNamespace(namespace);
-			List<MetricDatum> metricDatumList = new ArrayList<>();
-
-			// Iterating through the list of query results
-
-			for (Result result : results) {
-				try {
-					metricDatumList.add(processResult(result));
-				} catch (IllegalArgumentException iae) {
-					log.error("Could not convert result to double", iae);
+				PutMetricDataRequest metricDataRequest = new PutMetricDataRequest();
+				metricDataRequest.setNamespace(namespace);
+				List<MetricDatum> metricDatumList = new ArrayList<>();
+
+				// Iterating through the list of query results
+
+				for (Result result : results) {
+						try {
+								metricDatumList.add(processResult(result));
+						} catch (IllegalArgumentException iae) {
+								log.error("Could not convert result to double", iae);
+						}
 				}
-			}
 
-			metricDataRequest.setMetricData(metricDatumList);
-			cloudWatchClient.putMetricData(metricDataRequest);
+				metricDataRequest.setMetricData(metricDatumList);
+				cloudWatchClient.putMetricData(metricDataRequest);
 		}
 
 		private MetricDatum processResult(Result result) {
-			// Sometimes the attribute name and the key of the value are the same
-			MetricDatum metricDatum = new MetricDatum();
-			if (result.getValuePath().isEmpty()) {
-				metricDatum.setMetricName(result.getAttributeName());
-			} else {
-				metricDatum.setMetricName(result.getAttributeName() + "_" + KeyUtils.getValuePathString(result));
-			}
+				// Sometimes the attribute name and the key of the value are the same
+				MetricDatum metricDatum = new MetricDatum();
+				if (result.getValuePath().isEmpty()) {
+						metricDatum.setMetricName(result.getAttributeName());
+				} else {
+						metricDatum.setMetricName(result.getAttributeName() + "_" + KeyUtils.getValuePathString(result));
+				}
 
-			metricDatum.setDimensions(dimensions);
+				metricDatum.setDimensions(FluentIterable.from(this.dimensions)
+						.append(convertTypeNamesToDimensions(result.getTypeNameMap()))
+						.toList());
 
-			// Converts the Objects to Double-values for CloudWatch
-			metricDatum.setValue(toDoubleConverter.apply(result.getValue()));
-			metricDatum.setTimestamp(new Date());
-			return metricDatum;
+				// Converts the Objects to Double-values for CloudWatch
+				metricDatum.setValue(toDoubleConverter.apply(result.getValue()));
+				metricDatum.setTimestamp(new Date());
+				return metricDatum;
+		}
+
+		private ImmutableList<Dimension> convertTypeNamesToDimensions(final Map<String, String> typeNameMap) {
+				if (null != typeNames && null != typeNameMap && typeNameMap.size() > 0) {
+						return FluentIterable.from(typeNames)
+								.filter(new Predicate<String>() {
```

Comment:
Damn, we should upgrade to Java 8 and lamba at some point, this anonymous class hurts my eyes. I'm not sure how many users we still have on 7, but it has been end of life for long enough that we should probably not care. (this is just me ranting, don't do the Java 8 upgrade in this same PR).